osu.Game/Online/Chat/Channel.cs

```diff
@@ -38,16 +39,38 @@ public Channel()
         }
 
         public event Action<IEnumerable<Message>> NewMessagesArrived;
+        public event Action<IEnumerable<Message>> MessagesRemoved;
 
         public void AddNewMessages(params Message[] messages)
         {
             messages = messages.Except(Messages).ToArray();
 
-            Messages.AddRange(messages);
+            // A list of all tuples that contains all confirmed sent messages that are handled within this call
+            List<Tuple<LocalEchoMessage, Message>> handledConfirmedSentMessages = ConfirmedSentMessages.Where(messageTuple => messages.ToList().Contains(messageTuple.Item2)).ToList();
+
+            // Remove all confirmed sent messages that are handled within this call
+            foreach (Tuple<LocalEchoMessage, Message> handledTuple in handledConfirmedSentMessages)
+                Messages.Remove(handledTuple.Item1);
+
+            // Add local echo messages and error messages to the end of the list and received messages right in front of the first local echo message.
+            foreach (Message message in messages)
+            {
+                int insertionIndex = -1;
+
+                // Calculate the index of the first local echo message if the message is a received message
+                if (!(message is LocalEchoMessage) && !(message is ErrorMessage))
+                    insertionIndex = Messages.FindIndex(element => element is LocalEchoMessage);
+
+                // If there is no local echo message or if the message isn't a received message, insert the message at the list's very end
+                Messages.Insert(insertionIndex != -1 ? insertionIndex : Messages.Count, message);

```

Comment:
maybe just keep a counter `int pinnedMessageCount` and increment/decrement with the addition of error/echo messages?